src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,529 @@
+using System.Text;
+using System.Globalization;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using System.Threading;
+using System.Collections.Generic;
+using Microsoft.Cci;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// The string returned from this function represents the inputs to the compiler which impact determinism.  It is 
+    /// meant to be inline with the specification here:
+    /// 
+    ///     - https://github.com/dotnet/roslyn/blob/main/docs/compilers/Deterministic%20Inputs.md
+    /// 
+    /// Options which can cause compilation failure, but doesn't impact the result of a successful
+    /// compilation should be included. That is because it is interesting to describe error states
+    /// not just success states. Think about caching build failures as well as build successes.
+    ///
+    /// When an option is omitted, say if there is no value for a public crypto key, we should emit
+    /// the property with a null value vs. omitting the property. Either approach would produce 
+    /// correct results the preference is to be declarative that an option is omitted.
+    /// </summary>
+    internal abstract class DeterministicKeyBuilder
+    {
+        protected DeterministicKeyBuilder()
+        {
+        }
+
+        protected void WriteFilePath(
+            JsonWriter writer,
+            string propertyName,
+            string? filePath,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions options)
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
+            }
+            else if (filePath is not null)
+            {
+                filePath = PathUtilities.NormalizePathPrefix(filePath, pathMap);
+            }
+
+            writer.Write(propertyName, filePath);
+        }
+
+        internal static string EncodeByteArrayValue(ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            return builder.ToStringAndFree();
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
+        {
+            foreach (var b in value)
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
+
+        protected void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
```

Comment:
`Type?`?  can a null value actually sneak in?  feels... strange :)

Reply:
The majority of types used here are allowed to be `null`. I agree it's strange but it's allowed. 🤷 

Consider the concrete example of the `StrongNameProvider`

```c#
        /// <summary>
        /// Provides strong name and signature the source assembly.
        /// Null if assembly signing is not supported.
        /// </summary>
        public StrongNameProvider? StrongNameProvider { get; protected set; }
```